core/reflectx/value_of.go

```diff
@@ -58,34 +56,45 @@ func valueOf(s string, t reflect.Type) (reflect.Value, error) {
 	argInterface := argValue.Interface()
 
 	if json.Valid(argRaw) {
-		var err error
-		if protoMessage, ok := argInterface.(proto.Message); ok {
-			err = protojson.Unmarshal(argRaw, protoMessage)
-		} else {
-			err = json.Unmarshal(argRaw, argInterface)
-		}
-		if err == nil {
-			return outValue, nil
+		if err := json.Unmarshal(argRaw, argInterface); err != nil {
+			return outValue, errorValue(s, t, err)
 		}
-	}
 
-	if unmarshaler, ok := argInterface.(encoding.TextUnmarshaler); ok {
-		if err := unmarshaler.UnmarshalText(argRaw); err == nil {
-			return outValue, nil
-		}
+		return outValue, nil
 	}
 
-	if protoMessage, ok := argInterface.(proto.Message); ok {
-		if err := proto.Unmarshal(argRaw, protoMessage); err == nil {
-			return outValue, nil
+	if unmarshaler, ok := argInterface.(encoding.TextUnmarshaler); ok && utf8.ValidString(string(argRaw)) {
+		if err := unmarshaler.UnmarshalText(argRaw); err != nil {
+			return outValue, errorValue(s, t, err)
 		}
+
+		return outValue, nil
 	}
 
 	if unmarshaler, ok := argInterface.(encoding.BinaryUnmarshaler); ok {
-		if err := unmarshaler.UnmarshalBinary(argRaw); err == nil {
-			return outValue, nil
+		if err := unmarshaler.UnmarshalBinary(argRaw); err != nil {
+			return outValue, errorValue(s, t, err)
 		}
+
+		return outValue, nil
+	}
+
+	return outValue, errorValue(s, t, nil)
+}
+
+// errorValue constructs an error message for invalid argument value conversions.
+// Parameters:
+//   - arg: The argument value as a string.
+//   - t: The reflect.Type to which the argument was attempted to be converted.
+//   - errOrNil: The error encountered during the conversion, if any.
+//
+// Returns:
+//   - error: A formatted error message indicating the conversion failure.
+func errorValue(arg string, t reflect.Type, errOrNil error) error {
+	err := fmt.Errorf("%w: '%s': for type '%s'", ErrInvalidArgumentValue, arg, t.String())
+	if errOrNil != nil {
+		err = fmt.Errorf("%v: '%w'", err, errOrNil)
 	}
 
-	return outValue, fmt.Errorf("%w: '%s': for type '%s'", ErrInvalidArgumentValue, s, t.String())
+	return err
```

Comment:
I think it's supposed to be like this:

```
err := fmt.Errorf("%s: '%s' for type '%s': %w", ErrInvalidArgumentValue, "5", "int", errOrNil)
```

Reply:
Implemented wrapper. Probably it's not a best solution. But if you have any ideas about that - you're welcome.